### ISSUE TITLE
Service should be configurable globally

### ISSUE THREAD

**Issue by poteto** [poteto/ember-cli-flash#9]

**Comment by johnotander** [comment]
I'm wondering if you already have an approach you're considering for handling the global configuration for the service.

My initial idea for an approach was just to expand upon the `defaultX` properties like so:

```js
defaultTimeout() {
  return this._getGlobalDefaults().timeout || 3000;
},

// ... 

_getGlobalDefaults() {
  return Ember.ENV.flashMessageDefaults || {};
}
```

This has the benefit of working with the existing API (which I know is used by the flash object itself). But, it's a bit verbose, too. Is this similar to what you were intending @poteto? Or were you intending on going a different route?

**Comment by poteto** [comment]
I'm happy to rip out internal API for something better! I don't mind verbosity as long as it's understandable. Originally, I was going to do something like this, which is probably not the best idea:

```javascript
// pseudo
import config from '../../config/environment';

const defaults = {
  timeout      : 3000,
  priority     : 100,
  sticky       : false,
  showProgress : false,
  type         : 'info'
};

export default Ember.Service.extend({
  _setDefaults: on('init', function() {
    const { flashMessageDefaults: userDefaults } = config;

    Object.keys(defaults).map((key) => {
      const classifiedKey    = key.classify();
      const defaultKey       = `default${classifiedKey}`;
      const defaultValue     = defaults[key];
      const userDefaultValue = userDefaults[key];
      let valueToSet         = userDefaultValue ? userDefaultValue : defaultValue;

      set(this, defaultKey, valueToSet);
    });
  })
});
```

The idea is to still allow people to set those defaults directly on the service, e.g.

```javascript
get(this, 'flashes').set('defaultTimeout', 5000);
```

**Comment by johnotander** [comment]
Yeah, my proposed approach would be problematic because it wouldn't allow people to set the default directly. Observing `init` is a great approach. Though, I'd imagine we can merge `defaults` and `userDefaults` to clean up the loop?

```js
export default Ember.Service.extend({

  // ...

  _getDefaults() {
   const serviceDefaults = {
      timeout      : 3000,
      priority     : 100,
      sticky       : false,
      showProgress : false,
      type         : 'info'
    };

    const { flashMessageDefaults: userDefaults } = config;
    return Ember.$.extend({}, userDefaults, serviceDefaults);
  },

  _setDefaults: on('init', function() {
    var defaults = this._getDefaults();

    Object.keys(defaults).map((key) => {
      const classifiedKey = key.classify();
      const defaultKey    = `default${classifiedKey}`;

      set(this, defaultKey, defaults[key]);
    });
  })
});
```

**Comment by poteto** [comment]
Yeah, that's a lot cleaner. We can probably also use [`Ember.merge`](http://emberjs.com/api/#method_merge) instead of jQuery, I believe

**Comment by johnotander** [comment]
Oh sweet, TIL!

**Comment by poteto** [comment]
Teamwork!

![](http://media.giphy.com/media/pHb82xtBPfqEg/giphy.gif)

**Comment by poteto** [comment]
Closed in #36.
